pkg/eval/mods/path/path.go

```diff
@@ -176,3 +181,100 @@ func isRegular(path string) bool {
 	fi, err := os.Lstat(path)
 	return err == nil && fi.Mode().IsRegular()
 }
+
+//elvdoc:fn temp-dir
+//
+// ```elvish
+// temp-dir [&dir=$dir] [$pattern]
+// ```
+//
+// Create a temporary, unique, directory and output its name. The default for the `&dir` option is a
+// value appropriate for your system. You can use `&dir=/some/path` to specify a particular
+// directory in which to create the temp dir. The `$pattern` value is optional. If omitted it
+// defaults to `elvish-*`. The last star in the pattern is replaced by a random string. It is your
+// responsibility to remove the temporary directory.
+//
+// ```elvish-transcript
+// ~> path:temp-dir
+// ▶ /tmp/elvish-RANDOMSTR
+// ~> path:temp-dir x-
+// ▶ /tmp/x-RANDOMSTR
+// ~> path:temp-dir 'x-*.y'
+// ▶ /tmp/x-RANDOMSTR.y
+// ~> path:temp-dir &dir=.
+// ▶ elvish-RANDOMSTR
+// ~> path:temp-dir &dir=/some/dir
+// ▶ /some/dir/elvish-RANDOMSTR
+// ```
+
+type mktempOpt struct{ Dir string }
+
+func (o *mktempOpt) SetDefaultOptions() {}
+
+func tempDir(opts mktempOpt, args ...string) (string, error) {
+	var pattern string
+	switch len(args) {
+	case 0:
+		pattern = "elvish-*"
+	case 1:
+		pattern = args[0]
+	default:
+		return "", errs.ArityMismatch{
+			What:     "arguments here",
+			ValidLow: 0, ValidHigh: 1, Actual: len(args)}
+	}
```

Comment:
This is worth extracting to a function like:

```go
func getOptionalStringArg(args []string, default string) (string, error) { ... }
```

Reply:
FWIW, grepping the source for `len(args)` yielded only one other use that could use a `getOptionalStringArg` function: the [`cd` command](https://github.com/elves/elvish/blob/29034c20e9a30e9826c9f9aeac309b4d519db08e/pkg/eval/builtin_fn_fs.go#L69-L80). While my case has a zero cost default value the `cd` case does not. Which will require using a lambda, rather than string, arg for the default value. And even that is complicated by the need to handle an error return from the lambda that returns a default string.

I don't think a `getOptionalStringArg ` function is worth it at this time. This should be revisited if the ["here-values" syntax](https://github.com/elves/elvish/issues/1261) is implemented.

Reply:
I also looked at every file matched by `grep -h ' \.\.\.string)'` and, again, only found my two uses and the `cd` case. So I am still inclined to think the proposed abstraction is premature given that a default value cannot be counted on to be a static string.

Reply:
Also, it just occurred to me the default value might not even be a string. Which further complicates the implementation. Yes, the default value is a string for the one existing case, and the two cases I introduced, but that is by no means guaranteed to be true for future cases.

@xiaq, Feel free to comment further, and un-resolve this conversation, but as you can tell I feel pretty strongly that this is not the time to introduce that abstraction. Certainly not in the context of this pull-request which is already non-trivial.